File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.Random;

class DnDCharacter {
    private int strength, dexterity, constitution, intelligence, wisdom, charisma;
    private Random rnd;

    public DnDCharacter() {
        this.rnd = new Random();
        this.strength = ability();
        this.dexterity = ability();
        this.constitution = ability();
        this.intelligence = ability();
        this.wisdom = ability();
        this.charisma = ability();
    }

    int ability() {
        return rnd.ints(4, 1, 7).sorted().skip(1).sum();
    }

    int modifier(int constitution) {
        return Math.floorDiv(constitution - 10, 2);
    }

    int getStrength() {
        return this.strength;
    }

    int getDexterity() {
        return this.dexterity;
    }

    int getConstitution() {
        return this.constitution;
    }

    int getIntelligence() {
        return this.intelligence;
    }

    int getWisdom() {
        return this.wisdom;
    }

    int getCharisma() {
        return this.charisma;
    }

    int getHitpoints() {
        return 10 + modifier(this.constitution);
    }
}

File: solutions/go/dnd-character/2/dnd_character.go
package dndcharacter

import (
	"math"
	"math/rand"
	"slices"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

const (
	initialHitPoints  int = 10
	numberOfRolls     int = 4
	numberOfDiceSides int = 6
)

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return int(math.Floor((float64(score - 10.0)) / 2.0))
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	rolls := make([]int, 0, numberOfRolls)
	modifier := 0

	for i := 0; i < numberOfRolls; i++ {
		rolls = append(rolls, rand.Intn(numberOfDiceSides)+1)
		modifier += rolls[i]
	}

	modifier -= slices.Min(rolls)

	return modifier
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	consitution := Ability()
	return Character{
		Ability(),
		Ability(),
		consitution,
		Ability(),
		Ability(),
		Ability(),
		initialHitPoints + Modifier(consitution),
	}
}
